framework/src/org/checkerframework/framework/type/QualifierHierarchy.java

```diff
@@ -114,31 +114,41 @@ public abstract boolean isSubtype(
     public abstract AnnotationMirror leastUpperBound(AnnotationMirror a1, AnnotationMirror a2);
 
     /**
-     * Returns whether or not this hierarchy implements {@link #widenUpperBound(AnnotationMirror,
-     * AnnotationMirror)}
+     * Returns the number of iterations dataflow should preform before {@link
+     * #widenedUpperBound(AnnotationMirror, AnnotationMirror)} is called or -1 if it should never be
+     * called.
      *
-     * @return whether or not this hierarchy implements {@link #widenUpperBound(AnnotationMirror,
-     *     AnnotationMirror)}
+     * <p>Subclasses overriding this method should return some positive number.
+     *
+     * @return the number of iterations dataflow should preform before {@link
+     *     #widenedUpperBound(AnnotationMirror, AnnotationMirror)} is called or -1 if it should
+     *     never be called.
      */
-    public boolean implementsWidening() {
-        return false;
+    public int numberOfIterationsBeforeWidening() {
+        return -1;
     }
 
     /**
      * If the type hierarchy has an infinite ascending chain, then the dataflow analysis might never
      * reach a fixed point. To prevent this, implement this method such that it returns an upper
-     * bound for the two qualifiers that is wider than the least upper bound. If this method is
-     * implemented, also override {@link #implementsWidening()} and return true.
+     * bound for the two qualifiers that is a super type and not equal to the least upper bound. If
+     * this method is implemented, also override {@link #numberOfIterationsBeforeWidening()} and
+     * change its return to a positive number.
+     *
+     * <p>{@code newQualifier} is newest qualifier dataflow computed for some expression and {@code
+     * previousQualifier} is the qualifier dataflow computed on the last iteration.
      *
-     * <p>Otherwise, returns the least upper bound of the two annotations.
+     * <p>If the type hierarchy has no infinite ascending chain, returns the least upper bound of
+     * the two annotations.
      *
-     * @param a1 annotation
-     * @param a2 annotation
-     * @return an upper bound that is wider than the least upper bound of a1 and a2 (or the lub if
-     *     the type hierarchy does not require this)
+     * @param newQualifier newest qualifier dataflow computed for some expression
```

Comment:
I would use `new` and `previous` instead of `newest` and `last` to be symmetric with the parameter names.